### ISSUE TITLE
Potential memory leak after some long uptime?

### ISSUE THREAD

**Issue by Thomasedv** [Tribler/tribler#2966]
- [x] I have read CONTRIBUTING.rst

- [x] I have tried with the latest pre-release version and I still can reproduce the issue.

##### Tribler version/branch+revision: #####
Tribler RC2-2.

##### Operating system and version: #####
Windows 10

##### Expected behavior: #####
Tribler doesn't use GBs of memory. 

##### Actual behavior: #####
Came to my computer today, and was met with a low memory warning. Pointed to Tribler.
![image](https://cloud.githubusercontent.com/assets/1584461/26716346/588db18c-4779-11e7-8446-2ea59c01946c.png)
The large number in the third is the column for memory usage, if it wasn't immediately clear.  

##### Steps to reproduce the behavior: #####
Not sure, never happened before, and doesn't seem to happen immediately as far as i'm concerned. 

Edit: Forgot to mention, only one active download which wasn't progressing, Tribler closed fine when i tried, and the memory usage steadily increased with a few KB every few seconds, according to the task manger. 

##### Relevant log file output: #####
No mention of anything in the log.

**Comment by Thomasedv** [comment]
Update, started Tribler after this issue was made again, started around 150 MB memory used. Then steadily increased to 2GB used now, (6 Hours) Don't believe this was present in earlier versions of Tribler.

**Comment by synctext** [comment]
multiprocessing-fork ; very helpful!

**Comment by qstokkink** [comment]
Thanks for the report. In Tribler we have two processes: the GUI and the back end (of which the back end is the heaviest and should eat around 120 MB of memory). Both of them seem to have exploded in memory usage.
The way the memory freed itself suggests that this is some sort of wonky garbage collection. But this is just speculation.

**Comment by Thomasedv** [comment]
To bring an update, for now, there doesn't seem to be huge memory build ups anymore.(Well, unless what i mention below is true) I rebooted the PC, been a while, possibly a relation there. So, i'm monitoring Tribler. And mostly the memory usage seem to stay below 500 MB, for now at least. But there's a steady rise, and large drops in memory usage. Though, the lines are very tricky, because the scaling is changed on a by value basis to fit in, so be sure to look at the mentioned scaling in the second image. 

![image](https://cloud.githubusercontent.com/assets/1584461/26756775/c6fd8f7a-48aa-11e7-835f-fdf8fc170623.png)

On the bottom, the first value is average, the second minimum, and third is maximum. (in bytes) Last is duration, which is roughly one and a half hour. Take note on the black line, for what windows calls private bytes, which i think is the bytes used by Windows. 

The line that goes to the top if i'm correct, is the virtual memory. I have no clue really because one of the graphs is for "paging file" while another is for virtual memory usage(which should be the paging file for all i know). I'm not entirely sure, but i apparently reserved 15 GB on the HDD for that, so i'll see tomorrow if that one is full and the RAM suddenly starts going up without drops. Or if it's just some unrelated thing. 

Here's a full image with values:
![test](https://cloud.githubusercontent.com/assets/1584461/26756966/9c5430cc-48ae-11e7-8b76-a9c77142ef64.png)

Tribler is likely the GUI parts, and the multiprocess the Tribler# 1 They are the only ones i could monitor, so it's very likely it's those. 
From top to bot in listing: 
* "Private Working set"  (or something like that) Can see Tribler mentioned at the top, but Tribler# 1 is cut of because you can't resize that for some reason to see all of them. Tribler# 1 also got the same scaling as the others. 
* Byte in paging file.
* Byte in what google translate calls "interchangeable memory"
* Byte in virtual memory
* Private byte

I think these log files are shareable, and it's possible that the value names will show in English on a English OS(assuming you use Windows. Since this is built in), so if you want the logs, let me know.  Will let it go over night to see how things progress.

**Comment by devos50** [comment]
Thank you for your help getting more insight in this! Weird that after restarting your computer, the issue seems to be gone. It would be nice if we have a built-in memory profiler in the application itself and this might be something that we will build for Tribler 7.1.

**Comment by Thomasedv** [comment]
I'm not quite sure it's gone really. After looking letting it go over night, it's been at over 3 GB in use again. The interesting takeaway, seems that the accumulative effect seems to be something that's caused by Tribler not being actively used or checked. Or having active downloads. I started up a series of downloads over night, around midnight to 1pm. This morning in checked in, and Tribler had only one dl remaining, which is now done. Now look at the graph.

![current](https://cloud.githubusercontent.com/assets/1584461/26780093/65d0b65c-49e8-11e7-80b9-e7935b78e2b0.png)

Blank areas is when it was closed for some time. Take note how it increases over night, until in the morning. That's either because i actually used Tribler (checked it, interacted with it) or simply because my downloads finished. That's when you see that large drop off in the end. The black line is what's shown in task manager (90% sure it's that, it matches the memory used and the time i checked the computer at.) I'm a bit hesitant to say it's the download completion, because the last one was a bit slow, so it could have taken a bit longer than it took for Tribler to reduce the memory usage. Not really sure. 

Prior to the blank area, i don't think i had any active downloads, so it's more likely that that interaction with Tribler causes things to work "properly" again. That, or a more general case, where as long as the pc is locked or the user is AFK, for some reason, memory consumption rises. But short term, there's is a pattern here, that it does increase over night, when i don't use the computer.

**Comment by qstokkink** [comment]
I just performed a small replication test. My memory profiler did not show an increase in Python objects. However, I was greeted with this output in my terminal:

```
QObject::disconnect: Unexpected null parameter
QObject::disconnect: Unexpected null parameter
QObject::disconnect: Unexpected null parameter
QObject::disconnect: Unexpected null parameter
QObject::disconnect: Unexpected null parameter
QObject::disconnect: Unexpected null parameter
QObject::disconnect: Unexpected null parameter
QObject::disconnect: Unexpected null parameter
QObject::disconnect: Unexpected null parameter
```

This leads me to believe that Qt is piling up data while waiting for the screen to come back. I will test some more tomorrow to confirm this.

**Comment by qstokkink** [comment]
Small status update: we believe this (and other issues) to be caused by us distributing Qt (our GUI backend) version 5.2 with the Windows installer instead of the newest version.

**Comment by Thomasedv** [comment]
Can also provide a small update on my end. I update to Tribler rc2-12 (or 13, don't remember), and the issue was still there. (No fresh installation though) However, after my computer rebooted after a windows update, the issue seems far less apparent. The memory usage increase is much slower, and much less noticeable simply viewing the memory usage in task manager.  After almost 2 days, memory usage only reached 1GB, compared to above where it reached 4 GB over night. 

This does adds the possibility of it being a bug introduced in an update to Windows, and subsequently changed in another.

**Comment by qstokkink** [comment]
Confirmed on Ubuntu over the weekend, the GUI thread/process is up to 5.8 GB memory usage.

**Comment by devos50** [comment]
@qstokkink was that with Ubuntu 17.04? If so, this could be addressed to #3014.

Tomorrow, I will build functionality to export Meliae dumps from the GUI/core into the debug pane so we can easily debug memory information.

**Comment by qstokkink** [comment]
@devos50 no this was on the disposable laptop before we broke the internet during the Ubuntu 17 upgrade (I think it was running 16.04?).

**Comment by qstokkink** [comment]
Closing this issue as it is most likely fixed in its current form in #3094 .
